### ISSUE TITLE
New dependent: (JokeAPI) Full featured Java API

### ISSUE THREAD

**Issue by EasyG0ing1** [Sv443/JokeAPI#486]
## Link to the project:
[JokeAPI](https://github.com/EasyG0ing1/JokeAPI)
```
https://github.com/EasyG0ing1/JokeAPI
```
<br><br>

## Link to my personal website / GitHub page (optional):

```
https://github.com/EasyG0ing1
```
<br><br>

## Is your project explicit or unsafe for minors to view?

- [ ] **Yes**, it contains explicit material
- [X] **No**, it doesn't contain explicit material

<br><br>

## Notes:

This Java API includes every feature available from Sv443's API, including the ability to post new Jokes (though I have been getting a 500 `Internal Server Error` during testing so I have left that feature out of the documentation for now, though the code is written. Joke content searching, range definitions, all categories and black lists along with several different methods for users of the API to gain access to the jokes are included in this API. Even error reporting is included and accessible to users of the API, including all of the possible server response codes that can be seen in problematic situations.

<br><br>

## Checklist:
- [x] I have added my project's name to the title of this issue
- [x] I have filled out at least the first URL (second one is optional)
- [x] I have specified whether or not my project is explicit

**Comment by Sv443** [comment]
Also tell me if you have a unique name for it as I had to use the package identifier (or whatever it's called) to differentiate it from the other java wrapper

**Comment by EasyG0ing1** [comment]
@Sv443 I could rename the repository if that's what you're asking?

**Comment by Sv443** [comment]
> @Sv443 I could rename the repository if that's what you're asking? 

Only if you want to and you don't like what's currently in the readme

**Comment by Sv443** [comment]
> @Sv443 Also, do you know why I keep getting a 500 error from the server when I submit json data to it?
  
Yes, the submissions are closed until the next major release because I can't keep up with them with the current garbage system

**Comment by EasyG0ing1** [comment]
@Sv443 

> Yes, the submissions are closed until the next major release because I can't keep up with them with the current garbage system

This is good to know ... I'll change the library to indicate that submissions are not being accepted when it continues to return a 500 response yet leave the submission feature active for when the time comes that you're able to accept them again.

**Comment by Sv443** [comment]
> Is there any option of opening it up for dry runs only?
  
Good point, I was able to change it so dry runs are allowed now using the `?dry-run` parameter. The response code 201 indicates the format is correct, while 400 means it's not. The `additionalInfo` prop then tells you about which properties are missing or malformed.  
  
> I did enforce rules in the submission class that don't allow them to choose ANY as the type, and also it checks to make sure there is actual text in the right fields so that people can't inadvertently send annoying empty jokes, or they can't select a two-part joke while only filling out the joke field... the correct fields have to be filled out aligning with the chosen type before the class will send the joke.
  
That's aweseome!  
The next version would have a submission dashboard where people would be able to downvote bad submissions but it's always a good idea to pre-filter them.  
Also `Any` as a category isn't allowed in submissions and would give a 400 response.

**Comment by Sv443** [comment]
> I'm open to suggestions. I threw the repository and the readme together very quickly ... I'm not attached to any part of it emotionally lol
  
Sorry but if I was good at naming things this API wouldn't be called JokeAPI lol

**Comment by EasyG0ing1** [comment]
@Sv443 

> Sorry but if I was good at naming things this API wouldn't be called JokeAPI lol

Is that a joke?  🤣🤣🤣🤣

**Comment by EasyG0ing1** [comment]
@Sv443 

> Sorry but if I was good at naming things this API wouldn't be called JokeAPI lol

Perhaps `RandomJokeWithMultipleOptionsAPI` might have been more accurate ... 😉

**Comment by EasyG0ing1** [comment]
@Sv443 
> The response code 201 indicates the format is correct, while 400 means it's not. The `additionalInfo` prop then tells you about which properties are missing or malformed.

Is this `additionalInfo` prop just a string that you package into the server response along with the response code?

**Comment by EasyG0ing1** [comment]
@Sv443 

> The response code 201 indicates the format is correct, while 400 means it's not. The additionalInfo prop then tells you about which properties are missing or malformed.

I notice that the response comes back as a json object containing a boolean error, a String message and timestamp which is of course next to the actual response code separated by a semicolon...

My question is, if the response code ever comes back as 500 - I'm assuming that the json string will not be returned along with that response code as the response message in a 500 scenario would be much more generic in nature being a response from something other than the API itself, is that correct?

**Comment by Sv443** [comment]
> However, if the response code comes back as 400, there is no info as to what is missing from the joke, I only get the text
> 

I don't know what you're getting from the API but I'm getting this response:  
  
<img width="916" alt="image" src="https://github.com/Sv443/JokeAPI/assets/38158426/a3e4df9f-90f9-4a9f-ae3a-96646f65b6d9">  

<br><br>
  
> Also, if the response code ever comes back as 500 - I'm assuming that the json string will not be returned along with that response code as the response message in a 500 scenario would be much more generic in nature being a response from something other than the API itself, is that correct?

In the case of submissions and some other cases you would still get an error response in the same format, albeit maybe with more generic information and the `internalError` prop set to true.  
This is the main reason I always included the `error` prop and kept the error object format the same everywhere, so you have a quick way of knowing if there was an error and you don't have to do much validation.  
Though if the error code is *above* 500, such as 521 or 1015, you will probably be served an HTML error page from Cloudflare instead, so it might not be the worst idea to have a `Content-Type` header check for every response, or at least for errors.

**Comment by EasyG0ing1** [comment]
@Sv443 

> I don't know what you're getting from the API but I'm getting this response:

Can you show me the complete JSON String that you sent that returned that error?

**Comment by Sv443** [comment]
Sure:
```json
{
    "formatVersion": 3,
    "category": "Any",
    "type": "single",
    "joke": "A horse walks into a bar...",
    "flags": {
        "nsfw": true,
        "religious": false,
        "political": true,
        "racist": false,
        "sexist": false,
        "explicit": true
    },
    "lang": "en"
}
```
  
  
Changing the category to `Misc` returns a 201 with this object:
```json
{
    "error": false,
    "message": "Dry Run complete! No errors were found.",
    "timestamp": 1701770499213
}
```

**Comment by EasyG0ing1** [comment]
@Sv443 

When I send this:

```
{
  "formatVersion": 3,
  "category": "Any",
  "type": "single",
  "joke": "Why did the chicken cross the road? To get to the other side.",
  "flags": {
    "nsfw": false,
    "religious": false,
    "political": false,
    "racist": false,
    "sexist": false,
    "explicit": false
  },
  "lang": "en"
}
```

I get this back:
```
Server Code: 400
      Error: YES
    Message: Bad Request
  Timestamp: December 05, 2023, 02:00:36 AM
```

Of course I formatted the return from the server but that's the bits that I get

**Comment by Sv443** [comment]
Are you sure you're parsing the response message in case of an error? This only seems to be the basic HTTP info. There should still be some kind of request body you can read.

**Comment by EasyG0ing1** [comment]
@Sv443 

I figured it out, I had to check for a response code of `>=400` and if true, I had to grab the error stream from the connection so I could get the payload.

I'm kind of new to coding with web servers so its a learning process for me and this API has been good to learn from.

**Comment by Sv443** [comment]
Nice :)  
Honestly same here, this API was an entire learning experience, that's also why so many things are a bit weird about it, even to this day after I rewrote it from scratch for v2. I planned v3 originally to reiterate on my gained knowledge one last time and fix everything I don't like (which is quite a lot tbh).  
But in the end many people still like using it and it does work and so I'm happy about that.

**Comment by EasyG0ing1** [comment]
@Sv443 

> But in the end many people still like using it and it does work and so I'm happy about that.

You can't argue with results, and I see your API used a lot in example Arduino code which is what got me curious about it in the first place. Then when I saw that you actively look for wrappers written in other languages, I had to jump on it.

The only "cumbersome"-ness about your API - in my opinion, is the way that the SINGLE and TWOPART works ... its a bit awkward having to specifically deal with two different joke types. 

I think I would have opted instead, to have the joke itself sent and received with an ArrayList of Strings. That way, the joke itself can be one line or multiple lines as the joke submitter prefers. 

That would make the JSON of a numerous-line Joke look like this:

```JSON
{
  "id": 1678,
  "error": false,
  "category": "Misc",
  "lines": [
    "A priest, a pastor and a rabbi walk into a bar and soon begin arguing over who’s the best at what they do.",
    "",
    "Eventually they decide that in order to prove who’s the best, they would all go out alone into the woods and convert a bear to their respective religion.",
    "",
    "A few weeks later.. they meet up at the bar and the priest announces, \"I found a bear by the river and started talking to him about the Lord. He liked it so much that he now comes to mass every week.\"",
    "",
    "The pastor says, \"Well, I saw a bear in the clearing. I started reading him the bible and he loved it so much that he is now going to be baptized in about a week.\"",
    "",
    "The priest and the pastor turn to look at the rabbi, who now has a broken arm, a fractured collarbone and several cuts and bruises. The rabbi says, You know what, looking back.. maybe I shouldn\u0027t have started with circumcision"
  ],
  "flags": {
    "nsfw": false,
    "religious": true,
    "political": true,
    "racist": false,
    "sexist": false,
    "explicit": false
  },
  "safe": false,
  "lang": "en"
}
```

Or a two-liner would look like this:

```JSON
{
  "id": 1678,
  "error": false,
  "category": "Misc",
  "lines": [
    "Why did the chicken cross the road?",
    "To get to the other side."
  ],
  "flags": {
    "nsfw": false,
    "religious": false,
    "political": false,
    "racist": false,
    "sexist": false,
    "explicit": false
  },
  "safe": true,
  "lang": "en"
}
```

or a one-liner would look like this:
```JSON
{
  "id": 1678,
  "error": false,
  "category": "Misc",
  "lines": [
    "A recent double blind study done by the University of Chicago discovered a specific food that greatly reduces sex drive; wedding cake."
  ],
  "flags": {
    "nsfw": false,
    "religious": false,
    "political": false,
    "racist": false,
    "sexist": false,
    "explicit": false
  },
  "safe": true,
  "lang": "en"
}
```

The payload format is identical in all possible scenarios. This lets a joke be any size it needs to be while simplifying the handling of the joke at the same time.

**Comment by Sv443** [comment]
I get that this would make parsing simpler but at the same time I feel like there would still need to be some kind of separator that indicates where the "canonical parts" of the joke are, like in a Q&A style joke. I did it this way because it allows programmers to either concatenate them and treat them as a type=single joke or have them separated and only reveal the second part after the user clicks a button or after a timeout or something.  
So all in all I still think the current way is a good compromise.

**Comment by EasyG0ing1** [comment]
@Sv443 
I had not considered a scenario where there might be a desire to hide the response or time its reveal... makes sense under that scenario to do it this way.

Since that is the intended purpose of the two part joke, I'm going to modify the API a little and mention this use case in the README.

By the way, help me understand the need to rename my repository ... give me some context that helps me understand what your challenge is so I can pick a name that will work for you.

**Comment by Sv443** [comment]
> By the way, help me understand the need to rename my repository ... give me some context that helps me understand what your challenge is so I can pick a name that will work for you.

There really is no *need*, I just feel like a unique name could differentiate it better from the other Java wrapper lib but it's entirely up to you.  
  
![image](https://github.com/Sv443/JokeAPI/assets/38158426/e1b2bd7d-52c3-4847-8b5b-f355f4dae53e)

**Comment by EasyG0ing1** [comment]
> > By the way, help me understand the need to rename my repository ... give me some context that helps me understand what your challenge is so I can pick a name that will work for you.
> 
> There really is no _need_, I just feel like a unique name could differentiate it better from the other Java wrapper lib but it's entirely up to you.

I renamed it to **JavaJokesAPI**

`https://github.com/EasyG0ing1/JavaJokesAPI`

If you could maybe have it look like this?

![Screenshot 2023-12-05 at 9 33 36 AM](https://github.com/Sv443/JokeAPI/assets/10106834/61edd5e7-568c-4995-8bdc-a505979df587)

**Comment by EasyG0ing1** [comment]
@Sv443 I have a question about the data structure returned.

Heres some raw json that I get back:

```JSON
{
  "error": true,
  "internalError": false,
  "code": "107",
  "message": "Payload too large",
  "causedBy": [
    "The provided payload exceeds the limit of 5120 bytes (5.0 kB)."
  ],
  "timestamp": 1701798440441,
  "additionalInfo": "Payload too large (5762 Bytes - maximum is 5120 Bytes)."
}
```

and this one

```JSON
{
  "error": true,
  "internalError": false,
  "code": "105",
  "message": "Malformed Joke",
  "causedBy": [
    "This joke was formatted incorrectly."
  ],
  "timestamp": 1701798402996,
  "additionalInfo": "Submitted joke format is incorrect. Encountered error/s: Joke doesn\u0027t have a \"type\" property or it is invalid - it has to be either \"single\" or \"twopart\"."
}
```

```JSON
{
  "error": true,
  "internalError": true,
  "code": "100",
  "message": "Internal Error in HTTP Server",
  "causedBy": [
    "An error in the code - please contact me through one of the options on my website (https://sv443.net/) with the additional info."
  ],
  "timestamp": 1701799071425,
  "additionalInfo": "Joke submissions are disabled for the forseeable future."
}
```

Now that the actual server response codes that I get from these are either 400 or 500, but in the payloads , the "code" is in the 100's ... what does your code number signify that it should be different than the server response code?

**Comment by Sv443** [comment]
The API error codes are meant to be a quick way to differentiate error messages, either in error reports or on the clients too. They have nothing to do with HTTP and were just arbitrarily defined in here: https://github.com/Sv443/JokeAPI/blob/master/data/errorMessages.js  
Of all of them, I'd say only 100 and 104 are generic and would *need* a look at the `additionalInfo` prop

**Comment by Sv443** [comment]
I was able to modify the documentation page and your wrapper is added to it now.
